core/flow/tc_warm_up.go

```diff
@@ -0,0 +1,101 @@
+package flow
+
+import (
+	"math"
+	"sync/atomic"
+
+	"github.com/alibaba/sentinel-golang/core/base"
+	"github.com/alibaba/sentinel-golang/core/config"
+	"github.com/alibaba/sentinel-golang/util"
+)
+
+type WarmUpTrafficShapingCalculator struct {
+	threshold         float64
+	warmUpPeriodInSec uint32
+	coldFactor        uint32
+	warningToken      uint64
+	maxToken          uint64
+	slope             float64
+	storedTokens      *uint64
+	lastFilledTime    *uint64
+}
+
+func NewWarmUpTrafficShapingCalculator(warmUpPeriodInSec, warmUpColdFactor uint32, threshold float64) *WarmUpTrafficShapingCalculator {
```

Comment:
Maybe pass the * FlowRule is better? This way we don't need to change the function signature later if we need the rest of the rule's arguments